### ISSUE TITLE
Use a custom Http2HeadersDecoder to avoid copying

### ISSUE THREAD

**Issue by buchgr** [grpc/grpc-java#2217]
In the meeting today @carl-mastrangelo mentioned that reading headers from the wire and converting between metadata is a major contributor to our allocation profile. 

Can't we just use our own `Http2HeadersDecoder` that instead of [`DefaultHttp2Headers` ](https://github.com/netty/netty/blob/4.1/codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2HeadersDecoder.java#L82)uses our own `Http2Headers` implementation, which would just be an adapter for `Metadata` (wrap around `Metadata`)? We could thereby avoid copying between Metadata and `Http2Headers`, but HPACK would directly store to `Metadata`.

Am I missing something? 😅

**Comment by carl-mastrangelo** [comment]
@buchgr We want to reuse most of the HPACK decoder, but instead of getting an AsciiString out of the decoded value, we want to keep a strongly typed object instead the HPACK table.  (and also get proper eviction)   Is this possible today?

**Comment by buchgr** [comment]
There are two decoders: [`Http2HeadersDecoder`](https://github.com/netty/netty/blob/4.1/codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2HeadersDecoder.java) and [`Decoder`](https://github.com/netty/netty/blob/4.1/codec-http2/src/main/java/io/netty/handler/codec/http2/internal/hpack/Decoder.java). The first one is mostly a wrapper around the second one, which does the actual header decompression. I am talking about reusing the second one and effectively not having to call `Utils.convertHeaders(headers)`.

> but instead of getting an AsciiString out of the decoded value, we want to keep a strongly typed object instead the HPACK table.

I am not sure I understand. Do you want to have a String instead of AsciiString? Could you please elaborate?

**Comment by carl-mastrangelo** [comment]
One of the examples was an auth token of some kind: a higher order type than a string.   A Jwt might be sent as a header.  Rather than surface this as an AsciiString through the HPACK table, we could store the actual Jwt object in addition to the bytes representation.  That would save parsing it each time it is sent.  Likewise, when the jwt is evicted from the hpack table, the Jwt would also be removed at the same time.  If we did the mapping ourselves, we would have to keep track of the table state ourselves (as users of Netty, rather than inside of Netty.  

There are a lot of headers that don't change very often, but which we (grpc) reparse on each request.  I think we could save a lot by canonicalizing them.

**Comment by buchgr** [comment]
I have an idea that can give us all of that and I think we can get it merged into Netty.

The idea is to have the `Decoder` and `HuffmanDecoder` take something like a `CharSequenceFactory`. The default implementation returns`AsciiString` objects. These objects get cached in the dynamic table. That will be all the changes required in Netty.

gRPC will have its own `CharSequenceFactory`, that for keys returns `String` objects. For values it returns a special `CharSequence` that besides the binary data, can hold an additional object. Additionally, instead of passing the `DefaultHttp2Headers` to `Decoder.decode(ByteBuf, Http2Headers)`, we pass a custom `Http2Headers` implementation, that really just wraps around `Metadata`.

These changes will allows us to  to 1) cache higher kinded types than strings, 2) directly use `String` for keys, 3) avoid allocating `DefaultHttp2Headers` and 4) avoid converting between `Http2Headers` and `Metadata` i.e. `Utils.convertHeaders(headers)`.

The changes required to Netty are minimal and very generic. I think that should work. I ll throw together a PR for Netty + gRPC.

Thoughts?

**Comment by pgrosu** [comment]
@buchgr, @carl-mastrangelo I thought the meeting is tomorrow.  What meeting did I miss?

Thanks,
Paul

**Comment by ejona86** [comment]
The user forum meeting is tomorrow. The meeting mentioned was a Google-internal meeting where we coordinate the Google-grpc-java team, which also includes work on Stubby and gRPC within Google (which has proprietary goo added on top).

**Comment by pgrosu** [comment]
@ejona86 Thank you for clarifying.  Are there pertinent features that will trickle into gRPC?   I only ask since gRPC was advertised as the [new improved Stubby](http://www.grpc.io/blog/principles) that Google will transfer to also internally.

**Comment by ejona86** [comment]
@pgrosu, we do features in the open by default. We have to decide that a feature is Google-specific. That can be an obvious decision, like if it relies on Google-only infrastructure (say, server-side authentication). But even internally we are using the code unmodified (not strictly true, but modifications mainly deal with our internal build environment).

**Comment by pgrosu** [comment]
@ejona86 Super-duper-awesome!  That makes me really happy to hear, and yes that makes perfect sense now :)

**Comment by buchgr** [comment]
ping @carl-mastrangelo :-). Would love to get your input on that.

**Comment by carl-mastrangelo** [comment]
@buchgr The CharSequence that's a holder seems like it would work.  The devil is in the details so I would be curious how far you get with it.  I'm actually working on a Metadata improvement at the same time, so we should coordinate.

I have the change here: https://github.com/carl-mastrangelo/grpc-java/commit/84c2e3713d0d4ac65a8b25641120f6aec9b3ff04   if you want to see the easy changes I am doing in conjunction.

**Comment by buchgr** [comment]
@carl-mastrangelo 

Just had a look at your commit.

I have done effectively the same work for Netty's HTTP/2 headers a while back. It didn't make it into Netty, as it was slower if number of headers was large (IIRC 15+).

Some thoughts:

- gRPC Metadata should be small. So that could work.
- IIRC computing a hashCode on `put` and comparing that on `get`, instead of `Arrays.equals` was a big win.
- IIRC only the `char` overload of `Arrays.equals` is intrinsified. So for `byte[]` I implemented my own compare using `Unsafe`. I recently learned that Guava ships with a fast `byte[]` compare. So you could use that [1].
- Unrolling the loop helped not in microbenchmarks, but when running end-to-end benchmarks. IIRC that was because of cache misses. It's important to do the loads first. Intel CPUs have a line fill buffer with 10 entries, so they can do 10 cache misses in parallel.

```
for (...) {
  load[i]
  load[i+1] 
  load[i+2] 
  load[i+3]

  cmp [i]
  cmp [i+1]
  cmp [i+2]
  cmp [i+3]

  i += 4
}
```

**Comment by buchgr** [comment]
Hmm ... if my changes work out, we wouldn't need to do `Array.equals(asciiName(), names[i])`, but could use `String.equals`, which is much much faster (IIRC 4x of Arrays.equals). IIRC it's handwritten assembly, that uses vector instructions and so on :-).

**Comment by buchgr** [comment]
Man I hope our changes work out ... headers could be really really really fast. I am excited :-).

**Comment by carl-mastrangelo** [comment]
@buchgr Thanks for looking at the code.   There are some tradeoffs with it, but those are based on some ideas that aren't obvious from the commit:

* Ideally the bytes would not be compared _at all_.  I tried to have the byte arrays be able to be pointer comparisons, but it was tricky to make them that way off the wire.  There needs to be a fast was of only interning byte[] objects that have known keys.  
* With the changes to `GrpcHttp2Headers`, Metadata is almost completely empty.  The only serious header values left are accept-encoding, encoding, and timeout.  The size is always small.  Even custom application headers are rare.  
* We create new Metadata objects in a lot of places, but never user them.  The commit avoids allocating memory in that case.
* Keys and values are almost always read one, or written once.  Repeated use of headers is rare.

While making the code I noticed that the Metadata.Key is actually a convenient place to cache some of these values.  Maybe Keys can do the interning?   

Also, In your proposed change, were you thinking the Netty layer would reach into the Metadata object for the conversion to Http2Headers, or Metadata using a Netty provided "transport headers" maker?

**Comment by buchgr** [comment]
> Ideally the bytes would not be compared at all. I tried to have the byte arrays be able to be pointer comparisons, but it was tricky to make them that way off the wire. There needs to be a fast was of only interning byte[] objects that have known keys.

Not sure I understand. Could you elaborate please?

> With the changes to GrpcHttp2Headers, Metadata is almost completely empty. The only serious header values left are accept-encoding, encoding, and timeout. The size is always small. Even custom application headers are rare.

Do we need path, method and so on `Metadata` for incoming headers? I believe we have them right now, as we don't filter them, but I don't think we need them in there?

> Also, In your proposed change, were you thinking the Netty layer would reach into the Metadata object for the conversion to Http2Headers, or Metadata using a Netty provided "transport headers" maker?

Netty reaching into Metadata, I guess. Essentially `Http2Headers` wrapping around `Metadata`.

**Comment by carl-mastrangelo** [comment]
@buchgr When creating Metadata.Key, I try to make the asciiName normalize to a single byte[] object.  This is similar to String.intern() which ensures that you can use == to compare strings assuming that all of them are either literal strings, or interned.    This would be faster than looking at the bytes themselves I think.  This is easy for Keys, since we know them ahead of time, but harder for bytes read off the network.    Presumably when headers names are read off the wire, it could check to see if they equal any known keys and be interned().  

Good idea about the path and such, I don't think they need to be put in the metadata object.  I think the C core actually strips these headers out.

**Comment by buchgr** [comment]
@carl-mastrangelo 

That's what I meant with using `hashCode()`. With my proposed changes, we get that for free. The keys read of the network, will be String objects cached in HPACK's dynamic table. For a given key, all `Metadata` objects get the SAME String object that has its hash code precomputed. So looking up a key would be a simple integer comparsion for non-matching keys and a int comparison + equals for the matching key (`if (key0.hashCode() == key1.hashCode() && key0.equals(key1))`).

**Comment by buchgr** [comment]
Fixed by  8c18a0d
